helix-term/src/commands.rs

```diff
@@ -2206,6 +2240,7 @@ fn ensure_selections_forward(cx: &mut Context) {
 
 fn enter_insert_mode(cx: &mut Context) {
     cx.editor.mode = Mode::Insert;
+    cx.editor.search_position = None;
```

Comment:
I think insert-mode is not a good hook for clearing the search position:

* You can change the document in normal mode in a way that would affect the search result count; for example pasting or reloading the doc from disk
* This makes searching a kind of informal mode (like insert/normal/select) but it doesn't follow the normal behavior of clearing search results on Esc

Instead I think `Document::apply` would be a good place to clear this: that's when a transaction gets applied which might change the text in a document. That would also allow us to store and re-use the matches between successive calls of this function since we would be confident that if `doc.search_position.is_some()` then the document hasn't changed since the last search.

Reply:
done